api/middleware.go

```diff
@@ -0,0 +1,245 @@
+package api
+
+import (
+	"context"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/iancoleman/strcase"
+	apiTypes "github.com/oasisprotocol/oasis-indexer/api/v1/types"
+	"github.com/oasisprotocol/oasis-indexer/common"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/metrics"
+)
+
+var (
+	defaultOffset            = uint64(0)
+	defaultLimit             = uint64(100)
+	defaultBucketSizeSeconds = uint32(3600)
+	maxLimit                 = uint64(1000)
+)
+
+// MetricsMiddleware is a middleware that measures the start and end of each request,
+// as well as other useful request information.
+// It should be used as the outermost middleware, so it can
+// - set a requestID and make it available to all handlers and
+// - observe the final HTTP status code at the end of the request.
+func MetricsMiddleware(m metrics.RequestMetrics, logger log.Logger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Pre-work and initial logging.
+			requestID := uuid.New()
+			logger.Info("starting request",
+				"endpoint", r.URL.Path,
+				"request_id", requestID,
+			)
+			t := time.Now()
+			timer := m.RequestTimer(r.URL.Path)
+
+			// Serve the request.
+			next.ServeHTTP(w, r.WithContext(
+				context.WithValue(r.Context(), common.RequestIDContextKey, requestID),
+			))
+
+			// Observe results and log/record them.
+			httpStatus := reflect.ValueOf(w).Elem().FieldByName("status").Int()
+			logger.Info("ending request",
+				"endpoint", r.URL.Path,
+				"request_id", requestID,
+				"time", time.Since(t),
+				"status_code", httpStatus,
+			)
+			timer.ObserveDuration()
+
+			statusTxt := "failure"
+			if httpStatus >= 200 && httpStatus < 400 {
+				statusTxt = "success"
+			}
+			m.RequestCounter(r.URL.Path, statusTxt).Inc()
+		})
+	}
+}
+
+// ChainMiddleware is a middleware that adds chain-specific information
+// to the request context.
+func ChainMiddleware(chainID string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// snake_case the chainID; it's how it's used to name the DB schemas.
+			chainIDSnake := strcase.ToSnake(chainID)
+
+			// TODO: Set chainID based on provided height params.
+
+			next.ServeHTTP(w, r.WithContext(
+				context.WithValue(r.Context(), common.ChainIDContextKey, chainIDSnake),
+			))
+		})
+	}
+}
+
+// Sets a value for `Limit` and `Offset` fields of the given struct, if present and nil-valued.
+// oapi-codegen ignores the specified default value in the openapi spec:
```

Comment:
Yeah sucky of them. Anyway, we can impose defaults in this middleware (across all URLs) or in individual request handlers, which is what we were doing previously. So it's strictly better than what we had before ... but feels a little worse because codegen _should_ be handling it.